### DISCUSSION TITLE
segment和index相关问题

### DISCUSSION THREAD

**Discussion by RongHouchao** [milvus-io/milvus#20809] [Q&A and General discussion]
创建IVF_FLAT索引时，nlist参数是什么含义？是创建多少个簇吗？什么时候会起作用呢？查询的时候nprobe参数是什么意思？是查询的簇的个数吗？

向量数据构建索引时，每个segment会构建一个索引，那不同的segment之间的数据有没有构建索引？
segment的个数只跟数量量有关系吗？是不是达到segment存储数量量上限就会创建一个新的segment呢？
向量数据search时，queryNode会加载哪些segment？这个有没有加载机制？和查询参数nprobe有关系吗？

**Comment by yhmo** [comment]
nlist打个比方，有10000个人，我想找跟我最相像的那个人怎么找？如果一个一个比对过去的话，要比对10000次，这叫**暴搜**，这很低效。
那我先**按外貌相似**的人把他们分成若干组，比如分成50组，每组大约有200人（人数不是平均的）。然后每组里选出一个代表，我先和50个代表比对，比对nprobe次(nprobe <= nlist)，找到了nprobe个和我相似的代表，那他们所在的组里最有可能有和我最相似的人。于是我在这nprobe个组里和组里的人一个个比对。这样，比对了大约 50+nprobe*200 次后，就找到了跟我最相似的。
所以nlist就是分组(簇)的数量，nprobe就是找出多少个代表（我们称为中心点），在nprobe个簇中差找最终结果。如果nlist=50，nprobe=5的话，那搜索性能就是暴搜的10倍。

每个segment构建一个独立的索引。不同segment的索引相互之间没有任何关系。search在每个segment中搜索得到topk个结果，然后再和其他的segment的topk个结果合并，最终形成topk个结果返回给客户端。

segment的数量大致由segmentMaxSize（在milvus.yml的segment.maxSize定义）决定，默认512兆，就是说segment的大小大约是接近512兆这个数字，但具体的大小受milvus运行时insert/flush/compact等等行为的影响，产生几十兆的segment也是可能的。

search一个colletion时，queryode加载这个collection下面所属的所有segment，如果某个segment没有索引，则加载的是它的原始向量数据，搜索时对这个segment暴搜；如果某个segment已建好索引，则加载的是它的索引数据，搜索时用索引搜索。调用collection.load()则会加载这些数据。nprobe只是影响搜索时的行为，不会影响load的行为。

多看些文档啊兄弟，这些大多很基础，百度一下也行。

**Reply by RongHouchao** [reply]
好的，多谢，文档看了，写的没这么细，这些都是看了文档后产生的一些疑问。多谢解答了！
还有个疑问，search一个colletion时，queryode加载这个collection下面所属的所有segment，加入我有3个queryNode，3个queryNode加载的segment总和是全量数据，每个queryNode只加载其中的一部分？

假如说某个collection下数据量一共有10G数据，我有3台4G内存的机器组成集群，每台机器上有一个queryNode，那这样数据可以被正常加载并search吗？

**Reply by yhmo** [reply]
要看collection.load()时的参数replica设置是多少，默认是1，也就是每个segment在集群中只有一份放在内存里。如果replica=2，那么每个segment就有两份在内存里，milvus会把多份segemnt数据在各个querynode中做合理的分配。
假设用的replica是默认值1，那么3个querynode加载的segment总和就是这个collection的全量数据。
根据索引的类型不同和数据量不同，内存总量的需求也是不同。有个工具可以帮助估计：https://milvus.io/tools/sizing/

**Reply by RongHouchao** [reply]
ok，多谢。
但有个疑问，什么情况下会用到replica不是1的情况，多份segment在内存中有什么好处呢。我理解segment都是1份，在各个queryNode上search出topK数据，然后最终的结果进行聚合选出topK，这样不是最高效的么？类似MPP的架构

另外, replica=1时，在各个queryNode上search出topK数据，然后最终的结果进行聚合选出topK，这个聚合结果是在哪里？queryCoord上做的吗？还是proxy上做的？这一块我没太能串起来这个链路，能再帮忙解答一下吗？麻烦您了

**Reply by yhmo** [reply]
> ok，多谢。 但有个疑问，什么情况下会用到replica不是1的情况，多份segment在内存中有什么好处呢。我理解segment都是1份，在各个queryNode上search出topK数据，然后最终的结果进行聚合选出topK，这样不是最高效的么？类似MPP的架构
> 
> 另外, replica=1时，在各个queryNode上search出topK数据，然后最终的结果进行聚合选出topK，这个聚合结果是在哪里？queryCoord上做的吗？还是proxy上做的？这一块我没太能串起来这个链路，能再帮忙解答一下吗？麻烦您了

假设有2个segment，以及2个querynode，每个querynode处理一个segment。当某个querynode出问题不响应的时候，整个search服务就不能正常工作了，因为有一个segment的结果出不来。那在内存足够的前提下，我们把replica设为2，这样每个segment在内存都有两份，内部分配机制会把它们合理分配，这时，每个querynode上都有两个segment的数据，当某个querynode出问题了，那另一个querynode还能继续search这两个segment，服务不会中断。

topk合并的工作，先由querynode做第一阶段的合并，再由querycorrd做最终的合并。每个collection有shardnum，默认是2，每个shard会产生若干segment，querynode1就负责shard1的segments topk结果的合并，querynode2负责shard2的segments topk结果合并，最后在querycoord上把所有shard产生的topk做合并。

**Reply by RongHouchao** [reply]
ok，多谢，其实replica还承担着高可用的职责，可以这样理解吧。

这样会有个问题，如果有3个queryNode且replica=1，那我缩容的时候，把queryNode个数缩成了2个，那内存中就少了一部分数据了？这种情况由特殊处理机制？

”querynode1就负责shard1的segments topk结果的合并，querynode2负责shard2的segments topk结果合并“
----这里也有个疑问，
       如果有2个shard，3个querynode，那是怎么对应的？3个querynode中随机选2个处理2个shard的segments topk结果的合并？
      相反，如果有4个shard，2个querynode，是不是每个querynode分别处理2个shard的segments topk结果的合并？

**Reply by yhmo** [reply]
> ok，多谢，其实replica还承担着高可用的职责，可以这样理解吧。
> 
> 这样会有个问题，如果有3个queryNode且replica=1，那我缩容的时候，把queryNode个数缩成了2个，那内存中就少了一部分数据了？这种情况由特殊处理机制？
> 
> ”querynode1就负责shard1的segments topk结果的合并，querynode2负责shard2的segments topk结果合并“ ----这里也有个疑问， 如果有2个shard，3个querynode，那是怎么对应的？3个querynode中随机选2个处理2个shard的segments topk结果的合并？ 相反，如果有4个shard，2个querynode，是不是每个querynode分别处理2个shard的segments topk结果的合并？

把queryNode个数缩成了2，会触发内部的rebalance操作，于是有一个querynode上的segment就变成了2个

2个shard，3个querynode，那就有一个querynode在合并结果集的时候闲着，合并结果集不是一个耗时操作，并没有多大影响。4个shard，2个querynode，那就每个querynode各干两次。

**Reply by RongHouchao** [reply]
okok，多谢多谢，老哥还是稳啊，四平八稳！！！
